src/nmea.cpp

```diff
@@ -0,0 +1,1041 @@
+		9	The checksum data, always begins with *
+		*/
+
+		float track_true = 0.f;
+		char T;
+		float track_mag = 0.f;
+		char M;
+		float ground_speed = 0.f;
+		char N;
+		float ground_speed_K = 0.f;
+		char K;
+		NMEA_UNUSED(T);
+		NMEA_UNUSED(track_mag);
+		NMEA_UNUSED(M);
+		NMEA_UNUSED(N);
+		NMEA_UNUSED(ground_speed_K);
+		NMEA_UNUSED(K);
+
+		if (bufptr && *(++bufptr) != ',') {track_true = strtof(bufptr, &endp); bufptr = endp; }
+
+		if (bufptr && *(++bufptr) != ',') { T = *(bufptr++); }
+
+		if (bufptr && *(++bufptr) != ',') {track_mag = strtof(bufptr, &endp); bufptr = endp; }
+
+		if (bufptr && *(++bufptr) != ',') { M = *(bufptr++); }
+
+		if (bufptr && *(++bufptr) != ',') { ground_speed = strtof(bufptr, &endp); bufptr = endp; }
+
+		if (bufptr && *(++bufptr) != ',') { N = *(bufptr++); }
+
+		if (bufptr && *(++bufptr) != ',') { ground_speed_K = strtof(bufptr, &endp); bufptr = endp; }
+
+		if (bufptr && *(++bufptr) != ',') { K = *(bufptr++); }
+
+		float track_rad = track_true * M_PI_F / 180.0f; // rad in range [0, 2pi]
+
+		if (track_rad > M_PI_F) {
+			track_rad -= 2.f * M_PI_F; // rad in range [-pi, pi]
+		}
+
+		float velocity_ms = ground_speed / 1.9438445f;
+		float velocity_north = velocity_ms * cosf(track_rad);
+		float velocity_east  = velocity_ms * sinf(track_rad);
+
+		_gps_position->vel_m_s = velocity_ms;
+		_gps_position->vel_n_m_s = velocity_north;
+		_gps_position->vel_e_m_s = velocity_east;
+		_gps_position->cog_rad = track_rad;
+		_gps_position->vel_ned_valid = true; /** Flag to indicate if NED speed is valid */
+		_gps_position->c_variance_rad = 0.1f;
+		_gps_position->s_variance_m_s = 0;
+
+		if (!_VEL_received) {
+			_VEL_received = true;
+		}
+		// mavlink_log_info(&mavlink_log_pub, "get VTG data ");
+	}
+
+	if (_SVNUM_received &&
+	    _SVINFO_received &&
+	    _FIX_received
+	    ) {
+
+		_sat_num_gsv = _sat_num_gpgsv + _sat_num_glgsv + _sat_num_gagsv
+			      + _sat_num_gbgsv + _sat_num_bdgsv;
+
+		_gps_position->satellites_used = MAX(_sat_num_gns, _sat_num_gsv);
+		_gps_position->satellites_used = MAX(_gps_position->satellites_used, _sat_num_gga);
+	}
+
+	if (_VEL_received && _POS_received) {
+		ret = 1;
+		_gps_position->timestamp_time_relative = (int32_t)(_last_timestamp_time - _gps_position->timestamp);
+		_clock_set = false;
+		_VEL_received = false;
+		_POS_received = false;
+		_rate_count_vel++;
+		_rate_count_lat_lon++;
+	}
+
+	return ret;
+}
+
+int GPSDriverNMEA::receive(unsigned timeout)
+{
+	uint8_t buf[GPS_READ_BUFFER_SIZE];
+
+	/* timeout additional to poll */
+	uint64_t time_started = gps_absolute_time();
+
+	int j = 0;
+	int bytes_count = 0;
+
+	while (true) {
+
+		/* pass received bytes to the packet decoder */
+		while (j < bytes_count) {
+			int l = 0;
+
+			if ((l = parseChar(buf[j])) > 0) {
+				/* return to configure during configuration or to the gps driver during normal work
+				 * if a packet has arrived */
+				int ret = handleMessage(l);
+
+				if (ret > 0) {
+					return ret;
+				}
```

Comment:
Looking at this now, there could still be more data. The ubx driver handles this correctly.

Reply:
I'll look into the ubx driver then. I am not so confident fine-tuning these lower lever functions.

Reply:
@jajberni any update? I can also look into it if you want me to.

Reply:
Just in case this issue was already mentioned in previous discussion https://github.com/PX4/PX4-GPSDrivers/pull/58, so if it be usefull:

```
/* pass received bytes to the packet decoder */
while (_bytes_parsed < _bytes_readed) {
	int l = 0;

	if ((l = parseChar(_read_buf[_bytes_parsed])) > 0) {
		/* return to configure during configuration or to the gps driver during normal work
			* if a packet has arrived */
		if (handleMessage(l) > 0) {
			return 1;
		}
	}

	_bytes_parsed++;
}

/* everything is read */
_bytes_parsed = _bytes_readed = 0;
```

Reply:
Sorry, I have been sidetracked with other aspects of the project and haven't looked into this yet. Let me have a look and test it. If I don't find the time or any suitable solution, I will be more than happy to accept your help! :-)

Reply:
@jajberni I take it you didn't have time so I did the updates in https://github.com/PX4/PX4-GPSDrivers/tree/generic_nmea_driver, can you test it?